[BUG] Expected index does not match real index

### Describe the bug
Expected index does not match real index.

### To Reproduce
With checked executor:
```
df = pd.DataFrame(
            {
                "c1": [3, 4, 5, 3, 5, 4, 1, 2, 3],
                "c2": [1, 3, 4, 5, 6, 5, 4, 4, 4],
                "c3": [6, 5, 3, 9, 6, 8, 4, 3, 4],
            }
        )
mdf = md.DataFrame(df)
mdf = mdf.groupby("c1", as_index=False).count()["c2"]
```

1. Your Python version
2. The version of Mars you use
3. Versions of crucial packages, such as numpy, scipy and pandas
4. Full stack of the error.
```
  File "/Users/jon/Documents/repo/mars/mars/services/subtask/worker/tests/subtask_processor.py", line 84, in _execute_operand
    self.assert_object_consistent(out, ctx[out.key])
  File "/Users/jon/Documents/repo/mars/mars/tests/core.py", line 495, in assert_object_consistent
    self.assert_series_consistent(expected, real)
  File "/Users/jon/Documents/repo/mars/mars/tests/core.py", line 396, in assert_series_consistent
    self.assert_index_value_consistent(expected.index_value, real.index)
  File "/Users/jon/Documents/repo/mars/mars/tests/core.py", line 341, in assert_index_value_consistent
    raise AssertionError(
AssertionError: [address=127.0.0.1:64990, pid=42973] Index of real value (RangeIndex(start=0, stop=5, step=1)) not equal to (RangeIndex(start=0, stop=2, step=1))
```
5. Minimized code to reproduce the error.

### Expected behavior
A clear and concise description of what you expected to happen.

### Additional context
The problem is that  the index generated by `dataframe.groupby.core.build_mock_groupby` is set as the index_value of tileable's output.

--- a/mars/core/graph/core.pyx
+++ b/mars/core/graph/core.pyx
@@ -313,7 +313,12 @@ cdef class DirectedGraph:
     def _extract_operands(self, node):
         return [node.op]
 
-    def to_dot(self, graph_attrs=None, node_attrs=None, trunc_key=5, result_chunk_keys=None):
+    def to_dot(
+        self,
+        graph_attrs=None,
+        node_attrs=None,
+        trunc_key=5, result_chunk_keys=None, show_columns=False):
+
         sio = StringIO()
         sio.write('digraph {\n')
         sio.write('splines=curved\n')
@@ -330,6 +335,15 @@ cdef class DirectedGraph:
         operand_style = '[shape=circle]'
 
         visited = set()
+
+        def get_col_names(obj):
+            if hasattr(obj, "dtypes"):
+                return f"\"{','.join(list(obj.dtypes.index))}\""
+            elif hasattr(obj, "name"):
+                return f"\"{obj.name}\""
+            else:
+                return "\"N/A\""
+
         for node in self.iter_nodes():
             for op in self._extract_operands(node):
                 op_name = type(op).__name__
@@ -358,7 +372,10 @@ cdef class DirectedGraph:
                         sio.write(f'"{op_name}:{op.key[:trunc_key]}" {operand_style}\n')
                         visited.add(op.key)
                     sio.write(f'"{op_name}:{op.key[:trunc_key]}" -> '
-                              f'"Chunk:{self._gen_chunk_key(output_chunk, trunc_key)}"\n')
+                              f'"Chunk:{self._gen_chunk_key(output_chunk, trunc_key)}"')
+                    if show_columns:
+                        sio.write(f' [ label={get_col_names(output_chunk)} ]')
+                    sio.write("\n")
 
         sio.write('}')
         return sio.getvalue()
@@ -385,10 +402,10 @@ cdef class DirectedGraph:
 
         Fusion(self).decompose(nodes=nodes)
 
-    def view(self, filename='default', graph_attrs=None, node_attrs=None, result_chunk_keys=None):  # pragma: no cover
+    def view(self, filename='default', graph_attrs=None, node_attrs=None, result_chunk_keys=None, show_columns=False):  # pragma: no cover
         from graphviz import Source
 
-        g = Source(self.to_dot(graph_attrs, node_attrs, result_chunk_keys=result_chunk_keys))
+        g = Source(self.to_dot(graph_attrs, node_attrs, result_chunk_keys=result_chunk_keys, show_columns=show_columns))
         g.view(filename=filename, cleanup=True)
 
     def to_dag(self):

--- a/mars/dataframe/groupby/aggregation.py
+++ b/mars/dataframe/groupby/aggregation.py
@@ -235,7 +235,14 @@ class DataFrameGroupByAgg(DataFrameOperand, DataFrameOperandMixin):
         )
 
         shape = (np.nan, agg_df.shape[1])
-        index_value = parse_index(agg_df.index, groupby.key, groupby.index_value.key)
+        if isinstance(agg_df.index, pd.RangeIndex):
+            index_value = parse_index(
+                pd.RangeIndex(-1), groupby.key, groupby.index_value.key
+            )
+        else:
+            index_value = parse_index(
+                agg_df.index, groupby.key, groupby.index_value.key
+            )
 
         # make sure if as_index=False takes effect
         self._fix_as_index(agg_df.index)

--- a/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
+++ b/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
@@ -74,7 +74,10 @@ class ColumnPruningRule(OptimizationRule):
         if successor_required_columns is None:
             return None
         required_columns.update(successor_required_columns)
-        required_columns.update(self._get_self_required_columns(data))
+        self_required_columns = self._get_self_required_columns(data)
+        if self_required_columns is None:
+            return None
+        required_columns.update(self_required_columns)
         return required_columns
 
     @staticmethod

--- a/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
+++ b/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import List, Dict, Set, Any, Type, Union
+from typing import List, Dict, Set, Any, Type, Union, Optional
 
 import pandas as pd
 
@@ -68,9 +68,12 @@ class ColumnPruningRule(OptimizationRule):
     def _get_self_required_columns(self, data: TileableData) -> Set[Any]:
         return SelfColumnSelector.select(data)
 
-    def _get_required_columns(self, data: TileableData) -> Set[Any]:
+    def _get_required_columns(self, data: TileableData) -> Optional[Set[Any]]:
         required_columns = set()
-        required_columns.update(self._get_successor_required_columns(data))
+        successor_required_columns = self._get_successor_required_columns(data)
+        if successor_required_columns is None:
+            return None
+        required_columns.update(successor_required_columns)
         required_columns.update(self._get_self_required_columns(data))
         return required_columns
 
@@ -124,9 +127,11 @@ class ColumnPruningRule(OptimizationRule):
             op = data.op
 
             required_columns = self._get_required_columns(data)
-            if isinstance(op, ColumnPruneSupportedDataSourceMixin) and set(
-                required_columns
-            ) != self._get_all_columns(data):
+            if (
+                isinstance(op, ColumnPruneSupportedDataSourceMixin)
+                and required_columns is not None
+                and set(required_columns) != self._get_all_columns(data)
+            ):
                 op.set_pruned_columns(list(required_columns))
                 self.effective = True
                 pruned_nodes.append(data)
@@ -203,8 +208,10 @@ class ColumnPruningRule(OptimizationRule):
 
         for node in affected_nodes:
             required_columns = self._get_required_columns(node)
-            if isinstance(node, BaseDataFrameData) and set(required_columns) != set(
-                node.dtypes.index
+            if (
+                isinstance(node, BaseDataFrameData)
+                and required_columns is not None
+                and set(required_columns) != set(node.dtypes.index)
             ):
                 new_dtypes = pd.Series(
                     dict(

--- a/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
+++ b/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
@@ -17,6 +17,7 @@ from typing import List, Dict, Set, Any, Type, Union
 import pandas as pd
 
 from .input_column_selector import InputColumnSelector
+from .self_column_selector import SelfColumnSelector
 from ..core import register_optimization_rule
 from ...core import (
     OptimizationRecord,
@@ -52,9 +53,9 @@ class ColumnPruningRule(OptimizationRule):
         super().__init__(graph, records, optimizer_cls)
         self._context: Dict[TileableData, Dict[TileableData, Set[Any]]] = {}
 
-    def _get_selected_columns(self, data: TileableData) -> Set[Any]:
+    def _get_successor_required_columns(self, data: TileableData) -> Set[Any]:
         """
-        Get pruned columns of the given entity.
+        Get columns required by the successors of the given tileable data.
         """
         successors = self._get_successors(data)
         if successors:
@@ -64,12 +65,21 @@ class ColumnPruningRule(OptimizationRule):
         else:
             return self._get_all_columns(data)
 
+    def _get_self_required_columns(self, data: TileableData) -> Set[Any]:
+        return SelfColumnSelector.select(data)
+
+    def _get_required_columns(self, data: TileableData) -> Set[Any]:
+        required_columns = set()
+        required_columns.update(self._get_successor_required_columns(data))
+        required_columns.update(self._get_self_required_columns(data))
+        return required_columns
+
     @staticmethod
     def _get_all_columns(data: TileableData) -> Union[Set[Any], None]:
         """
-        Return all the columns of given tileable data. If the given tileable data is neither BaseDataFrameData nor
-        BaseSeriesData, None will be returned, indicating that column pruning is not available for the given tileable
-        data.
+        Return all the columns of given tileable data. If the given tileable data is neither
+        BaseDataFrameData nor BaseSeriesData, None will be returned, indicating that column pruning
+        is not available for the given tileable data.
         """
         if isinstance(data, BaseDataFrameData) and data.dtypes is not None:
             return set(data.dtypes.index)
@@ -82,8 +92,8 @@ class ColumnPruningRule(OptimizationRule):
         """
         Get successors of the given tileable data.
 
-        Column pruning is available only when every successor is available for column pruning (i.e. appears in the
-        context).
+        Column pruning is available only when every successor is available for column pruning
+        (i.e. appears in the context).
         """
         successors = list(self._graph.successors(data))
         if all(successor in self._context for successor in successors):
@@ -91,18 +101,18 @@ class ColumnPruningRule(OptimizationRule):
         else:
             return []
 
-    def _select_columns(self) -> None:
+    def _build_context(self) -> None:
         """
         Select required columns for each tileable data in the graph.
         """
         for data in self._graph.topological_iter(reverse=True):
             if self._is_skipped_type(data):
                 continue
-            self._context[data] = InputColumnSelector.select_input_columns(
-                data, self._get_selected_columns(data)
+            self._context[data] = InputColumnSelector.select(
+                data, self._get_required_columns(data)
             )
 
-    def _insert_getitem_nodes(self) -> List[TileableData]:
+    def _prune_columns(self) -> List[TileableData]:
         pruned_nodes: List[TileableData] = []
         datasource_nodes: List[TileableData] = []
 
@@ -112,12 +122,12 @@ class ColumnPruningRule(OptimizationRule):
                 continue
 
             op = data.op
-            selected_columns = self._get_selected_columns(data)
 
+            required_columns = self._get_required_columns(data)
             if isinstance(op, ColumnPruneSupportedDataSourceMixin) and set(
-                selected_columns
+                required_columns
             ) != self._get_all_columns(data):
-                op.set_pruned_columns(list(selected_columns))
+                op.set_pruned_columns(list(required_columns))
                 self.effective = True
                 pruned_nodes.append(data)
                 datasource_nodes.append(data)
@@ -192,15 +202,15 @@ class ColumnPruningRule(OptimizationRule):
                         affected_nodes.add(successor)
 
         for node in affected_nodes:
-            selected_columns = self._get_selected_columns(node)
-            if isinstance(node, BaseDataFrameData) and set(selected_columns) != set(
+            required_columns = self._get_required_columns(node)
+            if isinstance(node, BaseDataFrameData) and set(required_columns) != set(
                 node.dtypes.index
             ):
                 new_dtypes = pd.Series(
                     dict(
                         (col, dtype)
                         for col, dtype in node.dtypes.items()
-                        if col in selected_columns
+                        if col in required_columns
                     )
                 )
                 new_columns_value = parse_index(new_dtypes.index, store_data=True)
@@ -210,8 +220,8 @@ class ColumnPruningRule(OptimizationRule):
 
     @implements(OptimizationRule.apply)
     def apply(self):
-        self._select_columns()
-        pruned_nodes = self._insert_getitem_nodes()
+        self._build_context()
+        pruned_nodes = self._prune_columns()
         self._update_tileable_params(pruned_nodes)
 
     @staticmethod

--- a/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
+++ b/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
@@ -65,7 +65,8 @@ class ColumnPruningRule(OptimizationRule):
         else:
             return self._get_all_columns(data)
 
-    def _get_self_required_columns(self, data: TileableData) -> Set[Any]:
+    @staticmethod
+    def _get_self_required_columns(data: TileableData) -> Set[Any]:
         return SelfColumnSelector.select(data)
 
     def _get_required_columns(self, data: TileableData) -> Optional[Set[Any]]:

--- a/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
+++ b/mars/optimization/logical/tileable/column_pruning/column_pruning_rule.py
@@ -76,8 +76,6 @@ class ColumnPruningRule(OptimizationRule):
             return None
         required_columns.update(successor_required_columns)
         self_required_columns = self._get_self_required_columns(data)
-        if self_required_columns is None:
-            return None
         required_columns.update(self_required_columns)
         return required_columns
 
@@ -116,7 +114,7 @@ class ColumnPruningRule(OptimizationRule):
             if self._is_skipped_type(data):
                 continue
             self._context[data] = InputColumnSelector.select(
-                data, self._get_required_columns(data)
+                data, self._get_successor_required_columns(data)
             )
 
     def _prune_columns(self) -> List[TileableData]:
@@ -130,13 +128,13 @@ class ColumnPruningRule(OptimizationRule):
 
             op = data.op
 
-            required_columns = self._get_required_columns(data)
+            successor_required_columns = self._get_successor_required_columns(data)
             if (
                 isinstance(op, ColumnPruneSupportedDataSourceMixin)
-                and required_columns is not None
-                and set(required_columns) != self._get_all_columns(data)
+                and successor_required_columns is not None
+                and set(successor_required_columns) != self._get_all_columns(data)
             ):
-                op.set_pruned_columns(list(required_columns))
+                op.set_pruned_columns(list(successor_required_columns))
                 self.effective = True
                 pruned_nodes.append(data)
                 datasource_nodes.append(data)

--- a/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
@@ -62,7 +62,7 @@ class InputColumnSelector:
         cls,
         op_cls: OperandType,
         func: Callable[[TileableData, Set[Any]], Dict[TileableData, Set[Any]]],
-        replace: bool,
+        replace: bool = False,
     ) -> None:
         if op_cls not in cls._OP_TO_SELECT_FUNCTION or replace:
             cls._OP_TO_SELECT_FUNCTION[op_cls] = func

--- a/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
@@ -27,6 +27,7 @@ from .....dataframe.indexing.getitem import DataFrameIndex
 from .....dataframe.indexing.setitem import DataFrameSetitem
 from .....dataframe.merge import DataFrameMerge
 from .....typing import OperandType
+from .utils import get_cols_exclude_index
 
 
 class InputColumnSelector:
@@ -134,29 +135,19 @@ def df_merge_select_function(
                 suffix_col = str(col) + suffix
                 if suffix_col in required_cols:
                     ret[df].add(col)
+                    # The column in the other dataframe has to be selected as well. Otherwise, in
+                    # the runtime, there will not be a column with suffix.
+                    other_data = right_data if df is left_data else left_data
+                    ret[other_data].add(col)
 
     if op.on is not None:
-        ret[left_data].update(_get_cols_exclude_index(left_data, op.on))
-        ret[right_data].update(_get_cols_exclude_index(right_data, op.on))
+        ret[left_data].update(get_cols_exclude_index(left_data, op.on))
+        ret[right_data].update(get_cols_exclude_index(right_data, op.on))
     if op.left_on is not None:
-        ret[left_data].update(_get_cols_exclude_index(left_data, op.left_on))
+        ret[left_data].update(get_cols_exclude_index(left_data, op.left_on))
     if op.right_on is not None:
-        ret[right_data].update(_get_cols_exclude_index(right_data, op.right_on))
+        ret[right_data].update(get_cols_exclude_index(right_data, op.right_on))
 
-    return ret
-
-
-def _get_cols_exclude_index(inp: BaseDataFrameData, cols: Any) -> Set[Any]:
-    ret = set()
-    if isinstance(cols, (list, tuple)):
-        for col in cols:
-            if col in inp.dtypes.index:
-                # exclude index
-                ret.add(col)
-    else:
-        if cols in inp.dtypes.index:
-            # exclude index
-            ret.add(cols)
     return ret
 
 
@@ -184,7 +175,7 @@ def df_groupby_agg_select_function(
         selected_cols = set()
         # group by keys should be included
         if not groupby_series:
-            selected_cols.update(_get_cols_exclude_index(inp, by))
+            selected_cols.update(get_cols_exclude_index(inp, by))
         # add agg columns
         if op.raw_func is not None:
             if op.raw_func == "size":

--- a/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
@@ -62,7 +62,10 @@ class InputColumnSelector:
         op_cls: OperandType,
         func: Callable[[TileableData, Set[Any]], Dict[TileableData, Set[Any]]],
     ) -> None:
-        cls._OP_TO_SELECT_FUNCTION[op_cls] = func
+        if op_cls not in cls._OP_TO_SELECT_FUNCTION:
+            cls._OP_TO_SELECT_FUNCTION[op_cls] = func
+        else:
+            raise ValueError(f"key {op_cls} exists.")
 
     @classmethod
     def unregister(cls, op_cls: OperandType) -> None:
@@ -70,7 +73,7 @@ class InputColumnSelector:
             del cls._OP_TO_SELECT_FUNCTION[op_cls]
 
     @classmethod
-    def select_input_columns(
+    def select(
         cls, tileable_data: TileableData, required_cols: Set[Any]
     ) -> Dict[TileableData, Set[Any]]:
         """
@@ -238,7 +241,21 @@ def df_setitem_select_function(
         return ret
 
 
-SELECT_REQUIRED_OP_TYPES = [DataFrameBinOp, DataFrameUnaryOp, DataFrameIndex]
+@register_selector(DataFrameIndex)
+def df_getitem_select_function(
+    tileable_data: TileableData, required_cols: Set[Any]
+) -> Dict[TileableData, Set[Any]]:
+    if tileable_data.op.col_names:
+        return InputColumnSelector.select_required_input_columns(
+            tileable_data, required_cols
+        )
+    else:
+        return InputColumnSelector.select_all_input_columns(
+            tileable_data, required_cols
+        )
+
+
+SELECT_REQUIRED_OP_TYPES = [DataFrameBinOp, DataFrameUnaryOp]
 for op_type in SELECT_REQUIRED_OP_TYPES:
     InputColumnSelector.register(
         op_type, InputColumnSelector.select_required_input_columns

--- a/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
@@ -62,8 +62,9 @@ class InputColumnSelector:
         cls,
         op_cls: OperandType,
         func: Callable[[TileableData, Set[Any]], Dict[TileableData, Set[Any]]],
+        replace: bool,
     ) -> None:
-        if op_cls not in cls._OP_TO_SELECT_FUNCTION:
+        if op_cls not in cls._OP_TO_SELECT_FUNCTION or replace:
             cls._OP_TO_SELECT_FUNCTION[op_cls] = func
         else:
             raise ValueError(f"key {op_cls} exists.")

--- a/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/input_column_selector.py
@@ -132,6 +132,7 @@ def df_merge_select_function(
             if col in required_cols:
                 ret[df].add(col)
             else:
+                # TODO: this does not work when col is a tuple.
                 suffix_col = str(col) + suffix
                 if suffix_col in required_cols:
                     ret[df].add(col)

--- a/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
@@ -1,0 +1,77 @@
+# Copyright 2022-2023 XProbe Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Set, Any, Callable
+
+from mars.core import TileableData
+from mars.dataframe.indexing.getitem import DataFrameIndex
+from mars.dataframe.indexing.setitem import DataFrameSetitem
+from mars.typing import OperandType
+
+
+class SelfColumnSelector:
+
+    _OP_TO_SELECT_FUNCTION = {}
+
+    @classmethod
+    def register(
+        cls,
+        op_cls: OperandType,
+        func: Callable[[TileableData], Set[Any]],
+    ) -> None:
+        if op_cls not in cls._OP_TO_SELECT_FUNCTION:
+            cls._OP_TO_SELECT_FUNCTION[op_cls] = func
+        else:
+            raise ValueError(f"key {op_cls} exists.")
+
+    @classmethod
+    def select(cls, tileable_data: TileableData) -> Set[Any]:
+        """
+        TODO: docstring
+        """
+        op_type = type(tileable_data.op)
+        if op_type in cls._OP_TO_SELECT_FUNCTION:
+            return cls._OP_TO_SELECT_FUNCTION[op_type](tileable_data)
+        for op_cls in op_type.__mro__:
+            if op_cls in cls._OP_TO_SELECT_FUNCTION:
+                cls._OP_TO_SELECT_FUNCTION[op_type] = cls._OP_TO_SELECT_FUNCTION[op_cls]
+                return cls._OP_TO_SELECT_FUNCTION[op_cls](tileable_data)
+        return set()
+
+
+def register_selector(op_type: OperandType) -> Callable:
+    def wrap(selector_func: Callable):
+        SelfColumnSelector.register(op_type, selector_func)
+        return selector_func
+
+    return wrap
+
+
+@register_selector(DataFrameSetitem)
+def df_setitem_select_function(tileable_data: TileableData):
+    return {tileable_data.op.indexes}
+
+
+@register_selector(DataFrameIndex)
+def df_index_select_function(tileable_data: TileableData):
+    if tileable_data.op.col_names:
+        if isinstance(tileable_data.op.col_names, str):
+            return set([tileable_data.op.col_names])
+        else:
+            return set(tileable_data.op.col_names)
+    else:
+        return set(tileable_data.dtypes.index)
+
+
+# TODO: handle other ops

--- a/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
@@ -12,9 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Set, Any, Callable
+from typing import Set, Any, Callable, Iterable
 
 from mars.core import TileableData
+from mars.dataframe.core import BaseDataFrameData, BaseSeriesData
 from mars.dataframe.indexing.getitem import DataFrameIndex
 from mars.dataframe.indexing.setitem import DataFrameSetitem
 from mars.typing import OperandType
@@ -66,12 +67,16 @@ def df_setitem_select_function(tileable_data: TileableData):
 @register_selector(DataFrameIndex)
 def df_index_select_function(tileable_data: TileableData):
     if tileable_data.op.col_names:
-        if isinstance(tileable_data.op.col_names, str):
-            return set([tileable_data.op.col_names])
-        else:
+        col_names = tileable_data.op.col_names
+        if isinstance(col_names, Iterable):
             return set(tileable_data.op.col_names)
+        else:
+            return {tileable_data.op.col_names}
     else:
-        return set(tileable_data.dtypes.index)
+        if isinstance(tileable_data, BaseDataFrameData):
+            return set(tileable_data.dtypes.index)
+        elif isinstance(tileable_data, BaseSeriesData):
+            return {tileable_data.name}
 
 
 # TODO: handle other ops

--- a/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
@@ -25,7 +25,6 @@ from .....typing import OperandType
 
 
 class SelfColumnSelector:
-
     _OP_TO_SELECT_FUNCTION = {}
 
     @classmethod
@@ -33,7 +32,7 @@ class SelfColumnSelector:
         cls,
         op_cls: OperandType,
         func: Callable[[TileableData], Set[Any]],
-        replace: bool,
+        replace: bool = False,
     ) -> None:
         if op_cls not in cls._OP_TO_SELECT_FUNCTION or replace:
             cls._OP_TO_SELECT_FUNCTION[op_cls] = func

--- a/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Set, Any, Callable, Iterable
+from typing import Set, Any, Callable
 
 from .utils import get_cols_exclude_index
 from .....core import TileableData
@@ -74,7 +74,7 @@ def df_setitem_select_function(tileable_data: TileableData) -> Set[Any]:
 def df_getitem_select_function(tileable_data: TileableData) -> Set[Any]:
     if tileable_data.op.col_names:
         col_names = tileable_data.op.col_names
-        if isinstance(col_names, Iterable):
+        if isinstance(col_names, list):
             return set(tileable_data.op.col_names)
         else:
             return {tileable_data.op.col_names}
@@ -87,19 +87,51 @@ def df_getitem_select_function(tileable_data: TileableData) -> Set[Any]:
 
 @register_selector(DataFrameGroupByAgg)
 def df_groupby_agg_select_function(tileable_data: TileableData) -> Set[Any]:
+    """
+    Make sure the "group by columns" are preserved.
+    """
+
     op: DataFrameGroupByAgg = tileable_data.op
     by = op.groupby_params["by"]
 
     if isinstance(tileable_data, BaseDataFrameData):
         return get_cols_exclude_index(tileable_data, by)
     elif isinstance(tileable_data, BaseSeriesData):
-        return tileable_data.name
+        return {tileable_data.name}
     else:
         return set()
 
 
 @register_selector(DataFrameMerge)
 def df_merge_select_function(tileable_data: TileableData) -> Set[Any]:
+    """
+    Make sure the merge keys are preserved.
+    """
+
     op: DataFrameMerge = tileable_data.op
     on = op.on
-    return get_cols_exclude_index(tileable_data, on)
+    if on is not None:
+        return get_cols_exclude_index(tileable_data, on)
+
+    ret = set()
+    left_data: BaseDataFrameData = op.inputs[0]
+    right_data: BaseDataFrameData = op.inputs[1]
+    left_on = op.left_on
+    right_on = op.right_on
+    for data, merge_keys, suffix in zip(
+        [left_data, right_data], [left_on, right_on], op.suffixes
+    ):
+        for col in data.dtypes.index:
+            if col in merge_keys:
+                other_data = right_data if data is left_data else left_data
+                other_merge_keys = right_on if merge_keys is left_on else left_on
+
+                if col in other_data.dtypes.index and col not in other_merge_keys:
+                    # if the merge key exists in the other dataframe but not in the other
+                    # dataframe's merge keys, suffixes will be added.
+                    # TODO: this does not work when col is a tuple.
+                    suffix_col = str(col) + suffix
+                    ret.add(suffix_col)
+                else:
+                    ret.add(col)
+    return ret

--- a/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
@@ -116,11 +116,30 @@ def df_merge_select_function(tileable_data: TileableData) -> Set[Any]:
     ret = set()
     left_data: BaseDataFrameData = op.inputs[0]
     right_data: BaseDataFrameData = op.inputs[1]
-    left_on = op.left_on
-    right_on = op.right_on
+    left_index = op.left_index
+    right_index = op.right_index
+    left_on = op.left_on if isinstance(op.left_on, list) else [op.left_on]
+    right_on = op.right_on if isinstance(op.right_on, list) else [op.right_on]
+
+    if left_index and right_index:
+        return ret
+
+    if left_index:
+        for col in right_data.dtypes.index:
+            if col in right_on:
+                ret.add(col)
+        return ret
+    if right_index:
+        for col in left_data.dtypes.index:
+            if col in left_on:
+                ret.add(col)
+        return ret
+
     for data, merge_keys, suffix in zip(
         [left_data, right_data], [left_on, right_on], op.suffixes
     ):
+        if merge_keys is None:
+            continue
         for col in data.dtypes.index:
             if col in merge_keys:
                 other_data = right_data if data is left_data else left_data

--- a/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
@@ -33,8 +33,9 @@ class SelfColumnSelector:
         cls,
         op_cls: OperandType,
         func: Callable[[TileableData], Set[Any]],
+        replace: bool,
     ) -> None:
-        if op_cls not in cls._OP_TO_SELECT_FUNCTION:
+        if op_cls not in cls._OP_TO_SELECT_FUNCTION or replace:
             cls._OP_TO_SELECT_FUNCTION[op_cls] = func
         else:
             raise ValueError(f"key {op_cls} exists.")
@@ -72,7 +73,7 @@ def df_setitem_select_function(tileable_data: TileableData) -> Set[Any]:
 
 @register_selector(DataFrameIndex)
 def df_getitem_select_function(tileable_data: TileableData) -> Set[Any]:
-    if tileable_data.op.col_names:
+    if tileable_data.op.col_names is not None:
         col_names = tileable_data.op.col_names
         if isinstance(col_names, list):
             return set(tileable_data.op.col_names)

--- a/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
@@ -14,11 +14,14 @@
 
 from typing import Set, Any, Callable, Iterable
 
-from mars.core import TileableData
-from mars.dataframe.core import BaseDataFrameData, BaseSeriesData
-from mars.dataframe.indexing.getitem import DataFrameIndex
-from mars.dataframe.indexing.setitem import DataFrameSetitem
-from mars.typing import OperandType
+from .utils import get_cols_exclude_index
+from .....core import TileableData
+from .....dataframe.core import BaseDataFrameData, BaseSeriesData
+from .....dataframe.groupby.aggregation import DataFrameGroupByAgg
+from .....dataframe.indexing.getitem import DataFrameIndex
+from .....dataframe.indexing.setitem import DataFrameSetitem
+from .....dataframe.merge import DataFrameMerge
+from .....typing import OperandType
 
 
 class SelfColumnSelector:
@@ -60,12 +63,12 @@ def register_selector(op_type: OperandType) -> Callable:
 
 
 @register_selector(DataFrameSetitem)
-def df_setitem_select_function(tileable_data: TileableData):
+def df_setitem_select_function(tileable_data: TileableData) -> Set[Any]:
     return {tileable_data.op.indexes}
 
 
 @register_selector(DataFrameIndex)
-def df_index_select_function(tileable_data: TileableData):
+def df_getitem_select_function(tileable_data: TileableData) -> Set[Any]:
     if tileable_data.op.col_names:
         col_names = tileable_data.op.col_names
         if isinstance(col_names, Iterable):
@@ -79,4 +82,21 @@ def df_index_select_function(tileable_data: TileableData):
             return {tileable_data.name}
 
 
-# TODO: handle other ops
+@register_selector(DataFrameGroupByAgg)
+def df_groupby_agg_select_function(tileable_data: TileableData) -> Set[Any]:
+    op: DataFrameGroupByAgg = tileable_data.op
+    by = op.groupby_params["by"]
+
+    if isinstance(tileable_data, BaseDataFrameData):
+        return get_cols_exclude_index(tileable_data, by)
+    elif isinstance(tileable_data, BaseSeriesData):
+        return tileable_data.name
+    else:
+        return set()
+
+
+@register_selector(DataFrameMerge)
+def df_merge_select_function(tileable_data: TileableData) -> Set[Any]:
+    op: DataFrameMerge = tileable_data.op
+    on = op.on
+    return get_cols_exclude_index(tileable_data, on)

--- a/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/self_column_selector.py
@@ -64,7 +64,10 @@ def register_selector(op_type: OperandType) -> Callable:
 
 @register_selector(DataFrameSetitem)
 def df_setitem_select_function(tileable_data: TileableData) -> Set[Any]:
-    return {tileable_data.op.indexes}
+    if isinstance(tileable_data.op.indexes, list):
+        return set(tileable_data.op.indexes)
+    else:
+        return {tileable_data.op.indexes}
 
 
 @register_selector(DataFrameIndex)

--- a/mars/optimization/logical/tileable/column_pruning/tests/test_column_pruning.py
+++ b/mars/optimization/logical/tileable/column_pruning/tests/test_column_pruning.py
@@ -21,6 +21,7 @@ import pytest
 from ...... import dataframe as md
 from ...... import tensor as mt
 from ......dataframe.arithmetic import DataFrameMul
+from ......dataframe.base.eval import DataFrameEval
 from ......dataframe.base.isin import DataFrameIsin
 from ......dataframe.core import DataFrameData, SeriesData, DataFrameGroupByData
 from ......dataframe.datasource.dataframe import DataFrameDataSource
@@ -29,6 +30,7 @@ from ......dataframe.datasource.read_parquet import DataFrameReadParquet
 from ......dataframe.groupby.aggregation import DataFrameGroupByAgg
 from ......dataframe.groupby.core import DataFrameGroupByOperand
 from ......dataframe.indexing.getitem import DataFrameIndex
+from ......dataframe.indexing.setitem import DataFrameSetitem
 from ......dataframe.merge import DataFrameMerge
 from ......optimization.logical.tileable import optimize
 from ......tensor.core import TensorData
@@ -449,3 +451,142 @@ def test_two_groupby_aggs_with_multi_index(setup, gen_data2):
         if type(n.op) is DataFrameReadParquet and n.op.path == file_path
     ][0]
     assert read_parquet_node.op.get_columns() is None
+
+
+def test_merge_and_get_col_with_suffix(setup, gen_data1):
+    file_path, file_path2 = gen_data1
+    left = md.read_csv(file_path)
+    right = md.read_csv(file_path2)
+    r = left.merge(right, on="c1")[["c3_x"]]
+
+    graph = r.build_graph()
+    optimize(graph)
+
+    index_node = graph.result_tileables[0]
+    assert isinstance(index_node.op, DataFrameIndex)
+    assert index_node.op.col_names == ["c3_x"]
+
+    assert len(graph.predecessors(index_node)) == 1
+    merge_node = graph.predecessors(index_node)[0]
+    assert type(merge_node.op) is DataFrameMerge
+
+    read_csv_node_left, read_csv_node_right = graph.predecessors(merge_node)
+    assert type(read_csv_node_left.op) is DataFrameReadCSV
+    assert type(read_csv_node_right.op) is DataFrameReadCSV
+    assert len(read_csv_node_left.op.usecols) == 2
+    assert len(read_csv_node_right.op.usecols) == 2
+    assert set(read_csv_node_left.op.usecols) == {"c1", "c3"}
+    assert set(read_csv_node_right.op.usecols) == {"c1", "c3"}
+
+    r = r.execute().fetch()
+    raw1 = pd.read_csv(file_path)
+    raw2 = pd.read_csv(file_path2)
+    expected = raw1.merge(raw2, on="c1")[["c3_x"]]
+    pd.testing.assert_frame_equal(r, expected)
+
+
+def test_getitem_with_mask(setup, gen_data1):
+    """
+    Getitem with mask shouldn't prune any column.
+    """
+    file_path, file_path2 = gen_data1
+    df = md.read_csv(file_path)
+    df2 = md.read_csv(file_path2)
+
+    df = df[df2["c1"] > 3]
+    r = df.groupby(by="c1", as_index=False).sum()["c2"]
+
+    graph = r.build_graph()
+    optimize(graph)
+
+    index_node = graph.result_tileables[0]
+    assert isinstance(index_node.op, DataFrameIndex)
+    assert index_node.name == "c2"
+
+    assert len(graph.predecessors(index_node)) == 1
+    gb_node = graph.predecessors(index_node)[0]
+    assert isinstance(gb_node.op, DataFrameGroupByAgg)
+    assert set(gb_node.dtypes.index) == {"c1", "c2"}
+
+    assert len(graph.predecessors(gb_node)) == 1
+    index_node_2 = graph.predecessors(gb_node)[0]
+    isinstance(index_node_2.op, DataFrameIndex)
+    assert set(index_node_2.dtypes.index) == {"c1", "c2"}
+
+    assert len(graph.predecessors(index_node_2)) == 1
+    index_node_3 = graph.predecessors(index_node_2)[0]
+    isinstance(index_node_3.op, DataFrameIndex)
+    assert set(index_node_3.dtypes.index) == {"c1", "c2", "c3", "c4"}
+
+    assert len(graph.predecessors(index_node_3)) == 2
+    read_csv_node, eval_node = graph.predecessors(index_node_3)
+    assert isinstance(read_csv_node.op, DataFrameReadCSV)
+    assert isinstance(eval_node.op, DataFrameEval)
+    assert read_csv_node.op.usecols is None  # all the columns.
+    assert eval_node.name == "c1"
+
+    assert len(graph.predecessors(eval_node)) == 1
+    read_csv_node_2 = graph.predecessors(eval_node)[0]
+    assert isinstance(read_csv_node_2.op, DataFrameReadCSV)
+    assert read_csv_node_2.op.usecols == ["c1"]
+
+    raw1 = pd.read_csv(file_path)
+    raw2 = pd.read_csv(file_path2)
+    raw1 = raw1[raw2["c1"] > 3]
+    expected = raw1.groupby(by="c1", as_index=False).sum()["c2"]
+    pd.testing.assert_series_equal(
+        r.execute(extra_config={"check_series_name": False}).fetch(), expected
+    )
+
+
+def test_setitem(setup, gen_data1):
+    """
+    The output of DataFrameSetitem should preserve the column being set so that tile can work
+    correctly.
+    """
+    file_path, file_path2 = gen_data1
+    df = md.read_csv(file_path)
+    df2 = md.read_csv(file_path2)
+
+    df["c5"] = df2["c1"]
+    r = df.groupby(by="c1", as_index=False).sum()["c2"]
+
+    graph = r.build_graph()
+    optimize(graph)
+
+    index_node = graph.result_tileables[0]
+    assert isinstance(index_node.op, DataFrameIndex)
+    assert index_node.name == "c2"
+
+    assert len(graph.predecessors(index_node)) == 1
+    gb_node = graph.predecessors(index_node)[0]
+    assert isinstance(gb_node.op, DataFrameGroupByAgg)
+    assert set(gb_node.dtypes.index) == {"c1", "c2"}
+
+    assert len(graph.predecessors(gb_node)) == 1
+    index_node_2 = graph.predecessors(gb_node)[0]
+    isinstance(index_node_2.op, DataFrameIndex)
+    assert set(index_node_2.dtypes.index) == {"c1", "c2"}
+
+    assert len(graph.predecessors(index_node_2)) == 1
+    setitem_node = graph.predecessors(index_node_2)[0]
+    isinstance(setitem_node.op, DataFrameSetitem)
+    assert set(setitem_node.dtypes.index) == {"c1", "c2", "c5"}
+
+    assert len(graph.predecessors(setitem_node)) == 2
+    read_csv_node, index_node_3 = graph.predecessors(setitem_node)
+    assert isinstance(read_csv_node.op, DataFrameReadCSV)
+    assert isinstance(index_node_3.op, DataFrameIndex)
+    assert set(read_csv_node.op.usecols) == {"c1", "c2"}
+    assert index_node_3.name == "c1"
+
+    assert len(graph.predecessors(index_node_3)) == 1
+    read_csv_node_2 = graph.predecessors(index_node_3)[0]
+    assert isinstance(read_csv_node_2.op, DataFrameReadCSV)
+    assert read_csv_node_2.op.usecols == ["c1"]
+
+    raw1 = pd.read_csv(file_path)
+    raw2 = pd.read_csv(file_path2)
+    raw1["c5"] = raw2["c1"]
+    expected = raw1.groupby(by="c1", as_index=False).sum()["c2"]
+    pd.testing.assert_series_equal(r.execute().fetch(), expected)

--- a/mars/optimization/logical/tileable/column_pruning/tests/test_input_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/tests/test_input_column_selector.py
@@ -132,7 +132,7 @@ def test_df_merge():
     assert left.data in input_columns
     assert input_columns[left.data] == {"foo", 1}
     assert right.data in input_columns
-    assert input_columns[right.data] == {"foo"}
+    assert input_columns[right.data] == {"foo", 1}
 
     joined = left.merge(right, on=["foo", "bar"])
     input_columns = InputColumnSelector.select(joined.data, {"baz"})

--- a/mars/optimization/logical/tileable/column_pruning/tests/test_input_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/tests/test_input_column_selector.py
@@ -49,11 +49,11 @@ def test_register():
 
     InputColumnSelector.register(MockOperand, _select_input_columns)
     mock_data = MockEntityData()
-    assert InputColumnSelector.select_input_columns(mock_data, {"foo"}) == {}
+    assert InputColumnSelector.select(mock_data, {"foo"}) == {}
 
     # unregister
     InputColumnSelector.unregister(MockOperand)
-    assert InputColumnSelector.select_input_columns(mock_data, {"foo"}) == {
+    assert InputColumnSelector.select(mock_data, {"foo"}) == {
         MockOperand.get_mock_input(): {"foo", "bar"}
     }
 
@@ -69,31 +69,31 @@ def test_df_groupby_agg():
     )
 
     s = df.groupby(by="foo")["baz"].sum()
-    input_columns = InputColumnSelector.select_input_columns(s.data, {"baz"})
+    input_columns = InputColumnSelector.select(s.data, {"baz"})
     assert len(input_columns) == 1
     assert df.data in input_columns
     assert input_columns[df.data] == {"foo", "baz"}
 
     s = df.groupby(by=["foo", "bar"]).sum()
-    input_columns = InputColumnSelector.select_input_columns(s.data, {"baz"})
+    input_columns = InputColumnSelector.select(s.data, {"baz"})
     assert len(input_columns) == 1
     assert df.data in input_columns
     assert input_columns[df.data] == {"foo", "bar", "baz"}
 
     s = df.groupby(by="foo").agg(["sum", "max"])
-    input_columns = InputColumnSelector.select_input_columns(s.data, {"baz"})
+    input_columns = InputColumnSelector.select(s.data, {"baz"})
     assert len(input_columns) == 1
     assert df.data in input_columns
     assert input_columns[df.data] == {"foo", "baz"}
 
     s = df.groupby(by="foo")["bar", "baz"].agg(["sum", "max"])
-    input_columns = InputColumnSelector.select_input_columns(s.data, {"baz"})
+    input_columns = InputColumnSelector.select(s.data, {"baz"})
     assert len(input_columns) == 1
     assert df.data in input_columns
     assert input_columns[df.data] == {"foo", "bar", "baz"}
 
     s = df.groupby(by="foo").agg(new_bar=("bar", "sum"), new_baz=("baz", "sum"))
-    input_columns = InputColumnSelector.select_input_columns(s.data, {"new_bar"})
+    input_columns = InputColumnSelector.select(s.data, {"new_bar"})
     assert len(input_columns) == 1
     assert df.data in input_columns
     assert input_columns[df.data] == {"foo", "bar", "baz"}
@@ -104,7 +104,7 @@ def test_df_groupby_index_agg():
     df: DataFrame = DataFrame({"foo": (1, 1, 3), "bar": (4, 5, 6)})
     df = df.set_index("foo")
     s = df.groupby(by="foo").sum()
-    input_columns = InputColumnSelector.select_input_columns(s.data, {"bar"})
+    input_columns = InputColumnSelector.select(s.data, {"bar"})
     assert len(input_columns) == 1
     assert df.data in input_columns
     assert input_columns[df.data] == {"bar"}
@@ -116,39 +116,33 @@ def test_df_merge():
 
     joined = left.merge(right, on=["foo"])
 
-    input_columns = InputColumnSelector.select_input_columns(joined.data, {"foo"})
+    input_columns = InputColumnSelector.select(joined.data, {"foo"})
     assert left.data in input_columns
     assert input_columns[left.data] == {"foo"}
     assert right.data in input_columns
     assert input_columns[right.data] == {"foo"}
 
-    input_columns = InputColumnSelector.select_input_columns(
-        joined.data, {"foo", "baz"}
-    )
+    input_columns = InputColumnSelector.select(joined.data, {"foo", "baz"})
     assert left.data in input_columns
     assert input_columns[left.data] == {"foo"}
     assert right.data in input_columns
     assert input_columns[right.data] == {"foo", "baz"}
 
-    input_columns = InputColumnSelector.select_input_columns(
-        joined.data, {"foo", "1_x"}
-    )
+    input_columns = InputColumnSelector.select(joined.data, {"foo", "1_x"})
     assert left.data in input_columns
     assert input_columns[left.data] == {"foo", 1}
     assert right.data in input_columns
     assert input_columns[right.data] == {"foo"}
 
     joined = left.merge(right, on=["foo", "bar"])
-    input_columns = InputColumnSelector.select_input_columns(joined.data, {"baz"})
+    input_columns = InputColumnSelector.select(joined.data, {"baz"})
     assert left.data in input_columns
     assert input_columns[left.data] == {"foo", "bar"}
     assert right.data in input_columns
     assert input_columns[right.data] == {"foo", "bar", "baz"}
 
     joined = left.merge(right, on=["foo", "bar"])
-    input_columns = InputColumnSelector.select_input_columns(
-        joined.data, {"1_x", "1_y"}
-    )
+    input_columns = InputColumnSelector.select(joined.data, {"1_x", "1_y"})
     assert left.data in input_columns
     assert input_columns[left.data] == {"foo", "bar", 1}
     assert right.data in input_columns
@@ -163,7 +157,7 @@ def test_df_merge_on_index():
 
     # join on index
     joined = left.merge(right, on="foo")
-    input_columns = InputColumnSelector.select_input_columns(joined.data, {"baz"})
+    input_columns = InputColumnSelector.select(joined.data, {"baz"})
     assert left.data in input_columns
     assert input_columns[left.data] == set()
     assert right.data in input_columns
@@ -171,7 +165,7 @@ def test_df_merge_on_index():
 
     # left_on is an index and right_on is a column
     joined = left.merge(right, left_on="foo", right_on="bar")
-    input_columns = InputColumnSelector.select_input_columns(joined.data, {"baz"})
+    input_columns = InputColumnSelector.select(joined.data, {"baz"})
     assert left.data in input_columns
     assert input_columns[left.data] == set()
     assert right.data in input_columns
@@ -179,7 +173,7 @@ def test_df_merge_on_index():
 
     # left_on is a column and right_on is an index
     joined = left.merge(right, left_on="bar", right_on="foo")
-    input_columns = InputColumnSelector.select_input_columns(joined.data, {"baz"})
+    input_columns = InputColumnSelector.select(joined.data, {"baz"})
     assert left.data in input_columns
     assert input_columns[left.data] == {"bar"}
     assert right.data in input_columns
@@ -205,14 +199,14 @@ def test_df_arithmatic_ops():
 
     for op in ops:
         res: DataFrame = op(df1, 1)
-        input_columns = InputColumnSelector.select_input_columns(res.data, {"foo"})
+        input_columns = InputColumnSelector.select(res.data, {"foo"})
         assert len(input_columns) == 1
         assert res.data.inputs[0] in input_columns
         assert input_columns[res.data.inputs[0]] == {"foo"}
 
     for op in ops:
         res: DataFrame = op(df1, df2)
-        input_columns = InputColumnSelector.select_input_columns(res.data, {"foo"})
+        input_columns = InputColumnSelector.select(res.data, {"foo"})
         assert len(input_columns) == 2
         assert res.data.inputs[0] in input_columns
         assert input_columns[res.data.inputs[0]] == {"foo"}
@@ -232,21 +226,21 @@ def test_df_setitem():
 
     # scaler
     df[4] = 13
-    input_columns = InputColumnSelector.select_input_columns(df.data, {"foo"})
+    input_columns = InputColumnSelector.select(df.data, {"foo"})
     assert len(input_columns) == 1
     assert df.data.inputs[0] in input_columns
     assert input_columns[df.data.inputs[0]] == {"foo"}
 
     # scaler tensor
     df[5] = tensor()
-    input_columns = InputColumnSelector.select_input_columns(df.data, {"foo"})
+    input_columns = InputColumnSelector.select(df.data, {"foo"})
     assert len(input_columns) == 1
     assert df.data.inputs[0] in input_columns
     assert input_columns[df.data.inputs[0]] == {"foo"}
 
     # tensor
     df[6] = tensor([13, 14, 15, 16])
-    input_columns = InputColumnSelector.select_input_columns(df.data, {"foo"})
+    input_columns = InputColumnSelector.select(df.data, {"foo"})
     assert len(input_columns) == 2
     assert df.data.inputs[0] in input_columns
     assert input_columns[df.data.inputs[0]] == {"foo"}
@@ -255,7 +249,7 @@ def test_df_setitem():
 
     # series
     df[7] = Series([13, 14, 15, 16])
-    input_columns = InputColumnSelector.select_input_columns(df.data, {"foo"})
+    input_columns = InputColumnSelector.select(df.data, {"foo"})
     assert len(input_columns) == 2
     assert df.data.inputs[0] in input_columns
     assert input_columns[df.data.inputs[0]] == {"foo"}
@@ -264,7 +258,7 @@ def test_df_setitem():
 
     # dataframe
     df[[8, 9]] = df[["foo", "bar"]]
-    input_columns = InputColumnSelector.select_input_columns(df.data, {8})
+    input_columns = InputColumnSelector.select(df.data, {8})
     assert len(input_columns) == 2
     assert df.data.inputs[0] in input_columns
     assert input_columns[df.data.inputs[0]] == set()
@@ -282,7 +276,7 @@ def test_select_all():
         }
     )
     head = df.head()
-    input_columns = InputColumnSelector.select_input_columns(head.data, {"foo"})
+    input_columns = InputColumnSelector.select(head.data, {"foo"})
     assert len(input_columns) == 1
     assert head.data.inputs[0] in input_columns
     assert input_columns[head.data.inputs[0]] == {"foo", "bar", "baz", "qux"}

--- a/mars/optimization/logical/tileable/column_pruning/tests/test_input_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/tests/test_input_column_selector.py
@@ -280,3 +280,22 @@ def test_select_all():
     assert len(input_columns) == 1
     assert head.data.inputs[0] in input_columns
     assert input_columns[head.data.inputs[0]] == {"foo", "bar", "baz", "qux"}
+
+
+def test_getitem():
+    df: DataFrame = DataFrame(
+        {
+            "foo": (1, 1, 2, 2),
+            "bar": (3, 4, 3, 4),
+            "baz": (5, 6, 7, 8),
+            "qux": (9, 10, 11, 12),
+        }
+    )
+
+    getitem = df[df["foo"] == 1]
+    input_columns = InputColumnSelector.select(getitem.data, {"foo"})
+    assert input_columns[getitem.data.inputs[0]] == {"foo", "bar", "baz", "qux"}
+
+    getitem = df["foo"]
+    input_columns = InputColumnSelector.select(getitem.data, {"foo"})
+    assert input_columns[getitem.data.inputs[0]] == {"foo"}

--- a/mars/optimization/logical/tileable/column_pruning/tests/test_self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/tests/test_self_column_selector.py
@@ -64,6 +64,10 @@ def test_df_merge():
     required_columns = SelfColumnSelector.select(joined.data)
     assert required_columns == {"foo", "bar"}
 
+    joined = left.merge(right, left_on=["foo", "bar"], right_on=["foo", "bar"])
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo", "bar"}
+
     joined = left.merge(right)
     required_columns = SelfColumnSelector.select(joined.data)
     assert required_columns == {"foo", "bar", 1}
@@ -71,3 +75,27 @@ def test_df_merge():
     joined = left.merge(right, left_on=["foo"], right_on=["bar"])
     required_columns = SelfColumnSelector.select(joined.data)
     assert required_columns == {"foo_x", "bar_y"}
+
+    joined = left.merge(right, left_index=True, right_index=True)
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == set()
+
+    joined = left.merge(right, left_index=True, right_on="foo")
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo"}
+
+    joined = left.merge(right, left_index=True, right_on=["foo"])
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo"}
+
+    joined = left.merge(right, left_on="foo", right_index=True)
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo"}
+
+    joined = left.merge(right, left_on=["foo"], right_index=True)
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo"}
+
+    joined = left.merge(right, left_on=["foo"], right_index=True)
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo"}

--- a/mars/optimization/logical/tileable/column_pruning/tests/test_self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/tests/test_self_column_selector.py
@@ -95,7 +95,3 @@ def test_df_merge():
     joined = left.merge(right, left_on=["foo"], right_index=True)
     required_columns = SelfColumnSelector.select(joined.data)
     assert required_columns == {"foo"}
-
-    joined = left.merge(right, left_on=["foo"], right_index=True)
-    required_columns = SelfColumnSelector.select(joined.data)
-    assert required_columns == {"foo"}

--- a/mars/optimization/logical/tileable/column_pruning/tests/test_self_column_selector.py
+++ b/mars/optimization/logical/tileable/column_pruning/tests/test_self_column_selector.py
@@ -1,0 +1,73 @@
+# Copyright 2022-2023 XProbe Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ......dataframe import DataFrame
+from ..self_column_selector import SelfColumnSelector
+
+
+def test_df_setitem():
+    df = DataFrame({"foo": (1, 1, 3)})
+
+    df["bar"] = [1, 2, 3]
+    required_columns = SelfColumnSelector.select(df.data)
+    assert required_columns == {"bar"}
+
+    df[["baz", "qux"]] = 1, 2
+    required_columns = SelfColumnSelector.select(df.data)
+    assert required_columns == {"baz", "qux"}
+
+
+def test_df_getitem():
+    df = DataFrame({"foo": (1, 1, 3), "bar": (4, 5, 6)})
+
+    getitem = df["foo"]
+    required_columns = SelfColumnSelector.select(getitem.data)
+    assert required_columns == {"foo"}
+
+    getitem = df[["foo", "bar"]]
+    required_columns = SelfColumnSelector.select(getitem.data)
+    assert required_columns == {"foo", "bar"}
+
+
+def test_df_groupby_agg():
+    df = DataFrame({"foo": (1, 1, 3), "bar": (4, 5, 6)})
+
+    a = df.groupby(by="foo", as_index=False).sum()
+    required_columns = SelfColumnSelector.select(a.data)
+    assert required_columns == {"foo"}
+
+    a = df.groupby(by="foo").sum()
+    required_columns = SelfColumnSelector.select(a.data)
+    assert required_columns == set()
+
+
+def test_df_merge():
+    left = DataFrame({"foo": (1, 2, 3), "bar": (4, 5, 6), 1: (7, 8, 9)})
+    right = DataFrame({"foo": (1, 2), "bar": (4, 5), "baz": (5, 8), 1: (7, 8)})
+
+    joined = left.merge(right, on="foo")
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo"}
+
+    joined = left.merge(right, on=["foo", "bar"])
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo", "bar"}
+
+    joined = left.merge(right)
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo", "bar", 1}
+
+    joined = left.merge(right, left_on=["foo"], right_on=["bar"])
+    required_columns = SelfColumnSelector.select(joined.data)
+    assert required_columns == {"foo_x", "bar_y"}

--- a/mars/optimization/logical/tileable/column_pruning/utils.py
+++ b/mars/optimization/logical/tileable/column_pruning/utils.py
@@ -1,0 +1,31 @@
+# Copyright 2022-2023 XProbe Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Set
+
+from .....dataframe.core import BaseDataFrameData
+
+
+def get_cols_exclude_index(inp: BaseDataFrameData, cols: Any) -> Set[Any]:
+    ret = set()
+    if isinstance(cols, (list, tuple)):
+        for col in cols:
+            if col in inp.dtypes.index:
+                # exclude index
+                ret.add(col)
+    else:
+        if cols in inp.dtypes.index:
+            # exclude index
+            ret.add(cols)
+    return ret

